### ISSUE TITLE
Feature Request: Support Neptune-OSK

### ISSUE THREAD

**Issue by ACCESSDENIIED** [ayufan/steam-deck-tools#44]
Come across this post on reddit just now, the application allows us to use the OSK with both touchpads similar to SteamOS.

https://github.com/vilmire/neptune-osk

Is there a way to use this application alongside SDT? Possibly a way to disable input when Steam+X (Or OSK is displayed) then resume input when the OSK disappears?

Thanks :)

**Comment by ayufan** [comment]
@ACCESSDENIIED

I tested, but find it quite broken for me - as I don't see any difference. At least from perspective of this project you already have good OSK support:

- If you use Steam Controller, the `STEAM+X` already brings `Touch OSK`
- You can use `CTRL+WIN+O` as a shortcut triggered by `STEAM+X`
- Or just use `Steam OSK` that is part of a `-newbigpictureui`

It seems that some things are useful like styling OSK keyboard, and reworking how the mouse is used, but this appears that could be done without using Neptune Controller.

However, understanding how messy it is to run alongside Steam this tool is rather to cause more headaches than solves problems. I would rather say that this program could hook to generic system keyboard key press events, instead of trying to bind to Neptune Controller. The way how it does now, it messes with Lizard Buttons and Lizard Mouse settings which likely breaks interpretability with Steam, SWICD, Steam Controller or Glossi.

Given that I'm unsure if there's anything for me to integrate here. The only useful thing is styling OSK keyboard, but this is also something that will likely quickly fade away in favor of using Steam OSK that should be in a pretty good shape once the `-newbigpictureui` becomes stable.

cc @vilmire

**Comment by ayufan** [comment]
OK. Re-thinking this more. This CAN be nice, but I think `neptune-osk` should figure out how it can interoperate with other controller tools. Unsure what is the best way there:

- Maybe allow the `neptune-osk` to run on-demand, similar to how `tabtip.exe` works - start a process when OSK is requested. SteamController would call `neptune-osk` and configure `neptune-hid` accordingly (do not process in the interim the events from touchpads)
- Maybe make it hook to `ctrl-win-o` shortcut, and then figure out how to steal `neptune-controller`
- Maybe make `neptune-osk` to discover that user toggles `TouchOSK` and then "make it somehow connect to controller"

It should be very likely avoid touch `Lizard` settings. I also noticed that `neptune-osk.exe` is ~150MB - pretty crazy.

I think different applications reading `neptune-controller` is a problem.

**Comment by vilmire** [comment]
I made a pull request on SWICD and it doesn't touch Lizard.
Can steamdecktools do it too?
Oh and 150MB is the version with net6. can be 150 kb.

**Comment by ayufan** [comment]
> I made a pull request on SWICD and it doesn't touch Lizard.

I don't understand this. Can you expand what it does mean?

**Comment by ayufan** [comment]
@vilmire OK. I see what you do. Could you maybe prepare a single executable that when run would bring-up your OSK with Overlay? This should be easiest to manage, and then SDT could simply "detect it" and disable any touch pad interactions.

**Comment by vilmire** [comment]
@ayufan OK, I will make that version. So 
1. SDT will execute this exe. 
2. exe run osk.
3. play with lizard and touch input
4. exit when osk dead.
right?

**Comment by ayufan** [comment]
Hmm. Or maybe better. Basically `neptune-osk` could watch if OSK is shown (as person can also trigger OSK via notify icon). The `SDT` would discover that it is `neptune-osk` and would simply disable mousepad emulation to allow `neptune-osk` to temporarily take ownership of it.

I do like the idea of separate process.

**Comment by ayufan** [comment]
@vilmire 

I also find the `neptune-osk` not working for me on Win11 Pro.

I would also love to have a small executable instead of 150MB. You could expect that the `.net runtime` be installed for most, or provide a simple `install.bat` script alongside project to install `.net runtime` via `winget`. This is what I do as part of install: `winget install -e --id Microsoft.DotNet.DesktopRuntime.6 --override "/passive /norestart"`.

**Comment by vilmire** [comment]
> @vilmire
> 
> I also find the `neptune-osk` not working for me on Win11 Pro.
> 
> I would also love to have a small executable instead of 150MB. You could expect that the `.net runtime` be installed for most, or provide a simple `install.bat` script alongside project to install `.net runtime` via `winget`. This is what I do as part of install: `winget install -e --id Microsoft.DotNet.DesktopRuntime.6 --override "/passive /norestart"`.

thank you When I get home I'll figure out how to get it to work with win11pro and SDT.

**Comment by ayufan** [comment]
Try, I think we can find a good way to interwork together. I will make needed changes to discover `neptune-osk` when we figure out best way to interact.

**Comment by vilmire** [comment]
@ayufan 
https://github.com/vilmire/neptune-osk/releases/tag/v0.0.3
Just posted a new release. It turns on "NeptuneOskOverlay.exe" when the OSK opens and kills it when it closes.
I'll find out about win11pro over the weekend.

**Comment by vilmire** [comment]
I did a quick test for this. The result looks fine.

And I added an option to enable/disable Lizar in neptune-osk but it doesn't seem to work.

Settings that work is
- neptune-osk: emulate steam+x, enable, enable lizard mouse
- SDT: DoNotShow KeyboardStyle


this is just test code

Detect "NeptuneOskOverlay" in the DesktopProfile loop and stop emulating if found.
`
public override Status Run(Context c)
{
    if (base.Run(c).IsDone)
    {
        return Status.Done;
    }

    var process = Process.GetProcessesByName("NeptuneOskOverlay").FirstOrDefault();
    if (process is null)
    {
        if (!c.KeyboardMouseValid)
        {
            // Failed to acquire secure context
            // Enable emergency Lizard
            c.Steam.LizardButtons = true;
            c.Steam.LizardMouse = true;
        }
        else
        {
            c.Steam.LizardButtons = SettingsDebug.Default.LizardButtons;
            c.Steam.LizardMouse = SettingsDebug.Default.LizardMouse;
        }

        EmulateScrollOnLPad(c);
        EmulateScrollOnLStick(c);
        EmulateMouseOnRPad(c);
        EmulateMouseOnRStick(c);
        EmulateDPadArrows(c);

        c.Keyboard[VirtualKeyCode.RETURN] = c.Steam.BtnA;
        c.Keyboard[VirtualKeyCode.BACK] = c.Steam.BtnB;
    }
    else
    {
        c.Steam.LizardButtons = false;
        c.Steam.LizardMouse = false;
    }

    return Status.Continue;
}
`

**Comment by ayufan** [comment]
`c.Steam.LizardButtons = false;` I think this is to be removed.

**Comment by vilmire** [comment]
> > And I added an option to enable/disable Lizar in neptune-osk but it doesn't seem to work.
> 
> Don't do it. It is bad idea and bad design. Basically Steam and SDT and other controller software tries to manage, and if exposed to user it is confusing what it does. The problem with lizard setting is that it needs constantly be updated otherwise it will fallback to default.

Yes, I will change it so that users cannot manipulate it. I am not familiar with the lizard controller mechanism.

And
> `c.Steam.LizardButtons = false;` I think this is to be removed.

I'll remove it and test it.

**Comment by vilmire** [comment]
@ayufan 
remove lizardoption and remove LizardButton = false; Test done.  
https://github.com/vilmire/neptune-osk/releases/tag/v0.0.4

I will update Steamdeck for win Pro.

**Comment by vilmire** [comment]
I tested with winpro and it seems to work fine.(os build is 22621.963)
I am currently working on detaching overlay in netuneosk.
After detaching, the overlay does not depend on lizadcontroller. For now, neptune-osk send input. (STD can also send if you want)
And I'm about to do a pull request, but I don't know where is the best place to do process detection.

**Comment by ayufan** [comment]
I will look after tomorrow. Im currently out :)

**Comment by vilmire** [comment]
I ran into an issue while working on it. 
I confirmed that double-clicking occurs when running with NeptuneOsk and STD. 
To solve it, I commented out all the places related to the LizardMouse Enable or Disable in NeptuneOsk, but it works strangely. 
I think that the STD side should be managed together with the input, overlay. 
In the end, even if it's not one bundle, STD should run Overlay and send the controller input. 
Once the overlay is separated, I will comment again.

**Comment by ayufan** [comment]
Im currently out. Will check it by tomorrow :)

**Comment by vilmire** [comment]
The Overlay part (Touch visualize, Emulate) and the Winform part (Lizard function, OSK detection and execution) of neptune-osk are separated. 
Now we can detect the OSK on the STD, run it, and send the input to the Overlay to make it work. 
Shall I make a pull request about this?
It would be nice if it could be solved by process detection, but the lizard controller just doesn't seem to want more than one to exist. (double click issue)

**Comment by ayufan** [comment]
Sorry for the delay. Can you wait for my response for about one week? I
would love to see if I can finish the rest before giving you more advices.

**Comment by vilmire** [comment]
Okay. thank you.

**Comment by ACCESSDENIIED** [comment]
Any updates on this?
